include/dca/parallel/stdthread/thread_pool/thread_pool.hpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (C) 2012 Jakob Progsch, Václav Zeman
+// Copyright (C) 2018 ETH Zurich
+// Copyright (C) 2018 UT-Battelle, LLC
+// All rights reserved.
+//
+// See LICENSE for terms of usage.
+// See CITATION.md for citation guidelines, if DCA++ is used for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@itp.phys.ethz.ch)
+//
+// This file implements a thread pool based on the file
+// https://github.com/progschj/ThreadPool/blob/master/ThreadPool.h.
+// See COPYING for the license of the original work.
+
+#ifndef DCA_PARALLEL_STDTHREAD_THREAD_POOL_THREAD_POOL_HPP
+#define DCA_PARALLEL_STDTHREAD_THREAD_POOL_THREAD_POOL_HPP
+
+#include <condition_variable>
+#include <iostream>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <stdexcept>
+#include <thread>
+#include <vector>
+
+namespace dca {
+namespace parallel {
+
+class ThreadPool {
+public:
+  // Creates a pool with n_threads.
+  ThreadPool(size_t n_threads = 0);
+
+  ThreadPool(const ThreadPool& other) = delete;
+  ThreadPool(ThreadPool&& other) = default;
+
+  // Enlarge the pool to n_threads if the current number of threads in the pool is inferior.
+  void enlarge(std::size_t n_threads);
+
+  // Call asynchronously the function f with arguments args. This method is thread safe.
+  // Returns: a future to the result of f(args...).
+  template <class F, class... Args>
+  auto enqueue(F&& f, Args&&... args) -> std::future<typename std::result_of<F(Args...)>::type>;
+
+  // Conclude all the pending work and destroy the threds spawned by this class.
+  ~ThreadPool();
+
+  // Returns the number of threads used by this class.
+  std::size_t size() const {
+    return workers_.size();
+  }
+
+  // Returns a static instance.
+  static ThreadPool& get_instance() {
+    static ThreadPool global_pool;
+    return global_pool;
+  }
+
+private:
+  void workerLoop(int id);
+
+  // need to keep track of threads so we can join them
+  std::vector<std::thread> workers_;
+  // the task queue
+  std::vector<std::unique_ptr<std::queue<std::packaged_task<void()>>>> tasks_;
+
+  // synchronization
+  std::vector<std::unique_ptr<std::mutex>> queue_mutex_;
+  std::vector<std::unique_ptr<std::condition_variable>> condition_;
+  std::atomic<bool> stop_;
+  std::atomic<unsigned int> active_id_;
+};
+
+// add new work item to the pool
```

Comment:
@gbalduzz Please explain this in a line or two, with respect to what kind of work is getting in the pool. Is it a pool of threads or pool of work.

Reply:
Expanded in b05b9438e6b40e160423ef891f85d20c7bc30a1c